### PR TITLE
[GEP-7] Adds checks whether the target seed supports DNS before migration

### PR DESCRIPTION
**Pull request by plkokanov**
**How to categorize this PR?**

/area control-plane-migration
/kind enhancement
/priority normal

**What this PR does / why we need it**:
When the `Shoot.Spec.SeedName` field is updated to trigger migration, the current and new Seeds are checked whether they enable DNS or not. If either fo them (or both) does not enable DNS, the migration is forbidden.

**Which issue(s) this PR fixes**:
Part of #1631 

**Special notes for your reviewer**:
None

**Release note**:

```improvement operator
Migrating a Shoot from a Seed which has DNS disabled to a Seed which has DNS enabled will now generate a default domain name for the Shoot, if the Shoot's DNS section is not marked as `unmanaged`.
```
```improvement operator
Attempting to migrating a Shoot with `spec.dns` != nil  to a Seed which has DNS disabled is now forbidden and will return an error.
```

### CODE REVIEW

File: plugin/pkg/shoot/dns/admission.go [hunk 1]
```diff
@@ -196,7 +196,8 @@ func (d *DNS) Admit(ctx context.Context, a admission.Attributes, o admission.Obj
 		if shoot.Spec.SeedName == nil {
 			return nil
 		}
-		if oldShoot.Spec.SeedName != nil {
+
+		if oldShoot.Spec.SeedName != nil && *oldShoot.Spec.SeedName == *shoot.Spec.SeedName {
```
Comment (rfranzke) [inline plugin/pkg/shoot/dns/admission.go]: /invite @timuthy

**Review by rfranzke — COMMENTED** [review]
/lgtm

File: plugin/pkg/shoot/dns/admission.go (hunk 1 above)
Comment (rfranzke) [reply plugin/pkg/shoot/dns/admission.go]: ping @timuthy

**Review by timuthy — COMMENTED** [review]
The changes look good towards your description. 

> If the Shoot.Spec.DNS is nil and the new Seed supports DNS, then a default domain is generated for the Shoot (if the Shoot dns is not unmanaged)

However, I wonder if this is really the behavior we desire because it is irreversible. If a shoot from a non-DNS seed is migrated to a DNS seed, it'll never be able to run on a non-DNS seed again.  Opinions @plkokanov @rfranzke?

**Comment by rfranzke** [comment]
Good point @timuthy. I don't have a clear opinion yet. @plkokanov @swilen-iwanow what are your thoughts?

**Comment by swilen-iwanow** [comment]
Now when you point that out I would rather not go down that road, because it looks like a side effect of the migration

**Comment by plkokanov** [comment]
If I change this, should we completely forbid migration between seeds with different DNS settings or add handling for shoots without DNS running on seeds that support DNS (currently the reconciliation flow breaks with a nil ptr exception if that happens).

**Comment by timuthy** [comment]
> If I change this, should we completely forbid migration between seeds with different DNS settings or add handling for shoots without DNS running on seeds that support DNS (currently the reconciliation flow breaks with a nil ptr exception if that happens).

I vote for the the latter, i.e. let shoots w/o DNS run on DNS featured seeds.

**Comment by plkokanov** [comment]
@mvladev reminded me last week that if the Shoot does not have DNS, when it is migrated the APIServer will be recreated and a new IP can be assigned for it. This could lead to the APIServer (and other dependant) certs to have to be regenerated and the nodes on th shoot cluster restarted. I think we currently do not have logic to regenerate certificates in this way and restarting the nodes can be a huge burden.
I guess in the tests I did locally for the initial implementation I somehow got lucky.

So what I can do for now is just forbid migration of shoots that do not have DNS config. @rfranzke @timuthy @swilen-iwanow wdyt?

**Comment by rfranzke** [comment]
Yes, thanks, that's actually true, I forgot about it. I agree with forbidding migration of shoots scheduled to seeds with disabled DNS. If the shoot has an internal DNS record (but no external one) then it actually can be migrated (so just checking for shoots with `.spec.dns=nil` wouldn't be good enough).

**Review by rfranzke — DISMISSED** [review]
/lgtm
/invite @timuthy
/remove status/author-action
/unassign @plkokanov

**Comment by rfranzke** [comment]
ping @timuthy

File: plugin/pkg/shoot/dns/admission.go [hunk 2]
```diff
@@ -245,6 +252,24 @@ func checkFunctionlessDNSProviders(dns *core.DNS) error {
...
+			return apierrors.NewBadRequest(fmt.Sprintf("could not get referenced seed: %+v", err.Error()))
+		}
+		if seedDNSDisabled {
+			return apierrors.NewBadRequest("source and destination seeds must enable DNS so that the shoot can be migrated")
+		}
+	}
+
+	if newShoot.Spec.DNS != nil {
+		return checkFunctionlessDNSProviders(newShoot.Spec.DNS)
+	}
```
Comment (timuthy) [inline plugin/pkg/shoot/dns/admission.go]: I actually would like to refrain from calling this here because the admission code is already pretty nested, and responsibility wise `checkIfShootMigrationIsPossible` shouldn't call any provider validation.

**Review by timuthy — CHANGES_REQUESTED** [review]

File: plugin/pkg/shoot/dns/admission.go (hunk 2 above)
Comment (plkokanov) [reply plugin/pkg/shoot/dns/admission.go]: Just so you know why I called it here: 
Generally, if the seed is updated nothing prevents the user to also update the DNS Providers.
So when the seed is updated I wanted to check if migration is possible, check if the providers are ok and then always return from the function so that we don't continue and generate a default domain if the DNS of the shoot happens to be nil.

Comment (timuthy) [reply plugin/pkg/shoot/dns/admission.go]: Yes, I'm not objecting the use case but I think you can also check the DNS providers first, no matter if shoot migration is initiated or not and call `checkIfShootMigrationIsPossible` if necessary.

**Comment by plkokanov** [comment]
@timuthy could you have another look? Since we always have to return after checking the DNSProviders (otherwise the tests fail) I didn't move it at the top. Otherwise the code would look something like:
```
if oldShoot.Spec.SeedName != nil {
	if shoot.Spec.DNS != nil {
		if err := checkFunctionlessDNSProviders(shoot.Spec.DNS); err != nil {
			return err
		}
	}
	if *oldShoot.Spec.SeedName != *shoot.Spec.SeedName {
		return checkIfShootMigrationIsPossible(d.seedLister, oldShoot, shoot)
	}

	if shoot.Spec.DNS != nil {
		return nil
	}
}
```

**Review by timuthy — CHANGES_REQUESTED** [review]
Thanks @plkokanov for addressing my comment. Please see my suggestion [here](https://github.com/timuthy/gardener/commit/5c2a4234b499c9c074e23c9abae44d2465c155b0) to improve nesting and get rid of the duplicated invocation.

**Comment by plkokanov** [comment]
@timuthy adapted as you suggested. 
This is still a bit different than the original behaviour, as previously if the shoot.Spec.DNS == nil we would continue with the function. The tests do pass, but I'm not sure if there aren't any corner cases.

**Comment by timuthy** [comment]
> @timuthy adapted as you suggested.
> This is still a bit different than the original behaviour, as previously if the shoot.Spec.DNS == nil we would continue with the function. The tests do pass, but I'm not sure if there aren't any corner cases.

As discussed, you are right with that difference and even though it might not make sense to re-assign a removed `dns` section during updates, it's a breaking change we shouldn't introduce. Sorry I didn't get your point earlier.

**Comment by plkokanov** [comment]
Reverted to the previous change and added the additional test case as we discussed.

**Review by timuthy — APPROVED** [review]
/lgtm
Thanks @plkokanov

**Review by rfranzke — APPROVED** [review]
Thanks guys!
